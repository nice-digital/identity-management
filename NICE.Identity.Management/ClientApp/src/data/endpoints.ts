const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
