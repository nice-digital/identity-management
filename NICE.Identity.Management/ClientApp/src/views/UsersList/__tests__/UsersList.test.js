import React from "react";
import { act } from "react-dom/test-utils";
import { render, waitFor, waitForElementToBeRemoved, fireEvent, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter } from "react-router";
import { UsersList } from "../UsersList";
import { Endpoints } from "../../../data/endpoints";
import usersFour from "./usersFour.json";
import usersThirty from "./usersThirty.json";
import websites from "./websites.json";

const consoleErrorReset = console.error;

beforeEach(() => {
	console.error = consoleErrorReset;
});

const server = setupServer(
  rest.get(Endpoints.usersList, (req, res, ctx) => {
    return res(ctx.json(usersFour));
  }),
  rest.get(Endpoints.websitesList, (req, res, ctx) => {
    return res(ctx.json(websites));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("should show loading message before data has been loaded", () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "?amount=all&page=1",
		},
		history: {
			push: jest.fn(),
		},
	};
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	const loadingMessage = screen.getByText("Loading...", { selector: "p" });
	expect(loadingMessage).toBeInTheDocument();
});

test("should match the snapshot after data has been loaded", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "?amount=all",
		},
		history: {
			push: jest.fn(),
		},
	};
	const {container} = render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetch returns 401 error", async () => {
	console.error = jest.fn();	
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
		return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	const {container} = render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show error message when fetch returns 500 error", async () => {
	console.error = jest.fn();	
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	const {container} = render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show error message when fetch websites returns 401 error", async () => {
	console.error = jest.fn();	
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.websitesList, (req, res, ctx) => {
		return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	const {container} = render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show error message when fetch websites returns 500 error", async () => {
	console.error = jest.fn();	
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.websitesList, (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	const {container} = render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show no results message when fetch returns an empty array", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			return res.once(
				ctx.json([])
			);
		})
	);
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	const noResultsMessage = await screen.findByText("No results found", { selector: "p" });
	expect(noResultsMessage).toBeInTheDocument();
});


test("should show no results found message after search returns empty array", async () => {
	jest.useFakeTimers();
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	const dummyText = "SomeText";
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			const search = req.url.searchParams.get('q');
			
			if (search === dummyText) {
				return res.once(
					ctx.json([])
				);
			} else {
				return res(ctx.json(usersFour));				
			}
		})
	);
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	const searchInput = screen.getByLabelText("Filter by name or email address");
	searchInput.focus();
	await act(async () => {
		fireEvent.change(searchInput, {target: {value: dummyText}});
		jest.advanceTimersByTime(1000);
	});
	const noResultsFoundMessage = await screen.findByText(`No results found for "${dummyText}"`);
	expect(noResultsFoundMessage).toBeInTheDocument();
	jest.useRealTimers();
});

test("should filter users to all active when radio button is clicked", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const activeCheckbox = screen.getByLabelText("Active");
	const tagsBeforeFilter = screen.queryAllByText(/^(Active|Pending|Locked)/i, { selector: "span.tag"});
	expect(tagsBeforeFilter.length).toEqual(4);
	userEvent.click(activeCheckbox);
	await waitForElementToBeRemoved(() => screen.getAllByText("Locked", { selector: "span.tag" }));
	const tagsAfterFilter = screen.queryAllByText(/^(Active|Pending|Locked)/i, { selector: "span.tag"});
	const tagsAfterFilterCount = tagsAfterFilter.length;
	expect(tagsAfterFilterCount).toBeLessThan(4);
	tagsAfterFilter.forEach((tag) => {
		const {getByText} = within(tag);
		expect(getByText("Active", { selector: "span.tag"})).toBeInTheDocument();
	});
});

test("should filter users to those who have access to a specific website when radio button is clicked", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const eppiDevCheckbox = screen.getByTitle("Dev - EPPI Reviewer");
	const tagsBeforeFilterCount = screen.queryAllByText(/^(Active|Pending|Locked)/i, { selector: "span.tag"}).length;
	expect(tagsBeforeFilterCount).toEqual(4);
	userEvent.click(eppiDevCheckbox);
	await waitFor(() => screen.getByRole("heading", { name: "Showing 2 users" }));
	const tagsAfterFilterCount = screen.queryAllByText(/^(Active|Pending|Locked)/i, { selector: "span.tag"}).length;	
	expect(tagsAfterFilterCount).toBeLessThan(4);
});

test("should show 25 (default page amount) or less results by default when paginated", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			return res.once(
				ctx.json(usersThirty)
			);
		})
	);
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const usersList = screen.getByRole("list", { name: "users"});
	const { getAllByRole } = within(usersList);
	const usersListItems = getAllByRole("listitem");
	expect(usersListItems.length).toEqual(25);
});

test("should go to page 2 when next button is clicked", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			return res.once(
				ctx.json(usersThirty)
			);
		})
	);
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const nextPageButton = screen.getByRole("link", { name: "Go to next page"});
	userEvent.click(nextPageButton);
	const usersListSummary = await screen.findByText("Showing 26 to 30 of 30 users", { selector: "h2" });
	const paginationCounter = screen.getByText("Page 2 of 2");
	const usersList = screen.getByRole("list", { name: "users"});
	const { getAllByRole } = within(usersList);
	const usersListItems = getAllByRole("listitem");
	expect(usersListSummary).toBeInTheDocument();
	expect(paginationCounter).toBeInTheDocument();
	expect(usersListItems.length).toEqual(5);
});

test("should go to first page when page 1 button is clicked", async () => {
	const usersListProps = {
		location: {
			pathname: "/user",
			search: "?page=2",
		},
		history: {
			push: jest.fn(),
		},
	};
	server.use(
		rest.get(Endpoints.usersList, (req, res, ctx) => {
			return res.once(
				ctx.json(usersThirty)
			);
		})
	);
	render(<UsersList {...usersListProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const previousPageButton = screen.getByRole("link", { name: "Go to page 1"});
	userEvent.click(previousPageButton);
	const usersListSummary = await screen.findByText("Showing 1 to 25 of 30 users", { selector: "h2" });
	const paginationCounter = screen.getByText("Page 1 of 2");
	const usersList = screen.getByRole("list", { name: "users"});
	const { getAllByRole } = within(usersList);
	const usersListItems = getAllByRole("listitem");
	expect(usersListSummary).toBeInTheDocument();
	expect(paginationCounter).toBeInTheDocument();
	expect(usersListItems.length).toEqual(25);
});