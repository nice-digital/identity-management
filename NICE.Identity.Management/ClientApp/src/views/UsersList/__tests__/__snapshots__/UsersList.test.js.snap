// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList should match the snapshot after data has been loaded 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Users
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Users\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Users
    </h1>
  </div>,
  <div
    className="grid"
  >
    <div
      data-g="12 md:3"
    >
      <div
        className="panel"
      >
        <div
          className="input"
        >
          <label
            className="input__label"
            htmlFor="user-search"
          >
            Filter by name or email address
          </label>
          <input
            autoComplete="off"
            className="input__input"
            id="user-search"
            name="user-search"
            onChange={[Function]}
            type="search"
            unique="userSearch"
          />
        </div>
      </div>
      <div
        className="panel"
      >
        <form>
          <div
            aria-labelledby="filter-header"
            role="group"
          >
            <div
              id="filter-header"
            >
              <strong>
                Show:
              </strong>
            </div>
            <div
              className="radio"
            >
              <input
                className="radio__input"
                defaultChecked={true}
                id="filter-status-all"
                name="filter-status"
                onChange={[Function]}
                type="radio"
                value=""
              />
              <label
                className="radio__label"
                htmlFor="filter-status-all"
              >
                <span
                  className="visually-hidden"
                >
                  Show: 
                </span>
                All
              </label>
            </div>
            <div
              className="radio"
            >
              <input
                className="radio__input"
                id="filter-status-active"
                name="filter-status"
                onChange={[Function]}
                type="radio"
                value="active"
              />
              <label
                className="radio__label"
                htmlFor="filter-status-active"
              >
                Active
              </label>
            </div>
            <div
              className="radio"
            >
              <input
                className="radio__input"
                id="filter-status-pending"
                name="filter-status"
                onChange={[Function]}
                type="radio"
                value="pending"
              />
              <label
                className="radio__label"
                htmlFor="filter-status-pending"
              >
                Pending
              </label>
            </div>
            <div
              className="radio"
            >
              <input
                className="radio__input"
                id="filter-status-locked"
                name="filter-status"
                onChange={[Function]}
                type="radio"
                value="locked"
              />
              <label
                className="radio__label"
                htmlFor="filter-status-locked"
              >
                Locked
              </label>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      aria-busy={false}
      data-g="12 md:9"
    >
      <h2
        className="usersListSummary"
      >
        Showing 4 users
      </h2>
      <ul
        className="list--unstyled"
        data-qa-sel="list-of-users"
      >
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/users/1"
                    onClick={[Function]}
                  >
                    Kristin Patrick
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--consultation"
                      data-qa-sel="user-status"
                    >
                      Active
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Email address
                  </dt>
                  <dd>
                    kristin.patrick@example.com
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/users/2"
                    onClick={[Function]}
                  >
                    Thomas Holding
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--consultation"
                      data-qa-sel="user-status"
                    >
                      Active
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Email address
                  </dt>
                  <dd>
                    thomas.holding@example.com
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/users/3"
                    onClick={[Function]}
                  >
                    Aisha Bartlett
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--consultation"
                      data-qa-sel="user-status"
                    >
                      Active
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Email address
                  </dt>
                  <dd>
                    aisha.bartlett@example.com
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  <a
                    href="/users/4"
                    onClick={[Function]}
                  >
                    Theo Bradford
                  </a>
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag"
                      data-qa-sel="user-status"
                    >
                      Locked
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Email address
                  </dt>
                  <dd>
                    theo.bradford@example.com
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
      </ul>
      <div
        className="flex flexAlignCenter"
      >
        <div
          className="mr--e"
          data-qa-sel="number-of-results-on-page"
        >
          <label
            className="bold mr--c"
            htmlFor="itemsPerPage"
          >
            Items per page
          </label>
          <select
            data-qa-sel="result-on-the-page-index"
            id="itemsPerPage"
            name="itemsPerPage"
            onChange={[Function]}
            value="all"
          >
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="all"
            >
              All
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`UsersList should show error message when fetch returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Users
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Users\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Users
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`UsersList should show error message when fetch returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Users
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Users\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Users
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
